crates/subspace-transaction-pool/src/bundle_validator.rs

```diff
@@ -0,0 +1,256 @@
+use codec::Encode;
+use domain_runtime_primitives::Hash;
+use parking_lot::Mutex;
+use sc_client_api::{BlockBackend, HeaderBackend};
+use sp_api::{HeaderT, ProvideRuntimeApi};
+use sp_blockchain::HeaderMetadata;
+use sp_domains::{ExecutorApi, SignedOpaqueBundle};
+use sp_runtime::generic::BlockId;
+use sp_runtime::traits::{Block as BlockT, NumberFor, Zero};
+use std::collections::{HashSet, VecDeque};
+use std::sync::Arc;
+
+/// `BundleCollector` collect all the bundle from the last K (confirm depth) blocks
+/// of the best chain
+pub struct BundleCollector<Block: BlockT, Client> {
+    client: Arc<Client>,
+    // `bundle_syncer` used to sync `bundle_stored_in_last_k` to other thread
+    bundle_syncer: BundleStoredInLastK<Block>,
+    // The `(block_hashes, bundle_hashes)` of the last K blocks, sorted from newer block to
+    // older block
+    bundle_stored_in_last_k: VecDeque<(Block::Hash, HashSet<Hash>)>,
+    confirm_depth_k: usize,
+}
+
+impl<Block, Client> BundleCollector<Block, Client>
+where
+    Block: BlockT,
+    Client: HeaderBackend<Block>
+        + BlockBackend<Block>
+        + HeaderMetadata<Block, Error = sp_blockchain::Error>
+        + ProvideRuntimeApi<Block>,
+    Client::Api: ExecutorApi<Block, Hash>,
+{
+    pub fn new(client: Arc<Client>, confirm_depth_k: usize) -> Self {
+        BundleCollector {
+            client,
+            bundle_syncer: BundleStoredInLastK::new(),
+            bundle_stored_in_last_k: VecDeque::new(),
+            confirm_depth_k,
+        }
+    }
+
+    pub fn bundle_validator(&self) -> BundleValidator<Block, Client> {
+        BundleValidator {
+            client: self.client.clone(),
+            bundle_syncer: self.bundle_syncer.clone(),
+        }
+    }
+
+    /// Get the hash of the current best block, return `None` if it have not process
+    /// any block.
+    fn best_hash(&self) -> Option<Block::Hash> {
+        self.bundle_stored_in_last_k
+            .front()
+            .map(|(block_hash, _)| *block_hash)
+    }
+
+    fn collect_bundle_from_block(
+        &self,
+        block_hash: Block::Hash,
+    ) -> Result<(Block::Hash, HashSet<Hash>), sp_blockchain::Error> {
+        let bundle_hashes: HashSet<_> = self
+            .client
+            .runtime_api()
+            .extract_stored_bundle_hashes(&BlockId::Hash(block_hash))?
+            .into_iter()
+            .collect();
+        Ok((block_hash, bundle_hashes))
+    }
+
+    fn collect_bundle_from_last_k(
+        &mut self,
+        mut hash: Block::Hash,
+    ) -> Result<(), sp_blockchain::Error> {
+        assert!(self.bundle_stored_in_last_k.is_empty());
```

Comment:
An error is usually preferred.

Reply:
Error is unnecessary, this is kind of a defense (like the `expect`, and `panic`), the caller to this private function will check `bundle_stored_in_last_k` is empty before calling it.

Reply:
Then a message is better to be added to explain why it's okay.

Reply:
We can not, the `bundle_stored_in_last_k` is storing the bundle hash from the last **consecutive** K blocks of the same fork, if `bundle_stored_in_last_k` is not empty calling this function may cause its state mixup (i.e non-consecutive). 

Also, this function is for initialization and it should only call once when `bundle_stored_in_last_k` is empty. (it will be renamed to `initialize_recent_stored_bundles` as @liuchengxu suggested)